### PR TITLE
ObjectLoader: Convert to es6 class.

### PR DESCRIPTION
**Pull request by mrdoob**
This should allow geometries to be tree-shacked?

See #19986 (comment).

### CODE REVIEW

**Comment by marcofugaro** [comment]
@mrdoob ugh, testing this branch, `ObjectLoader` isn't being three-shaken because [it's modified in place here](https://github.com/mrdoob/three.js/blob/c6ab22b7c518ba6e427696499d026eb2962e0162/src/Three.Legacy.js#L471-L480), in the legacy files.

I really don't know how we should handle legacy methods, adding a method outside the class is modifying the class in place. Any idea @donmccurdy?

-------------

By commenting these lines, ~30kb more get shaved off of the final bundle, it's most of the unused geometries.

Screenshot 2020-09-01 at 23 48 36

**Comment by mrdoob** [comment]
Ah, we can simply move that method to the class.

**Comment by marcofugaro** [comment]
Nice! I think it's more tidy

**Comment by ianpurvis** [comment]
Funny- I was also testing `ObjectLoader` last night... I found that you can get it to shake without the es6 class?

```js
ObjectLoader.prototype = /*@__PURE__*/ Object.assign( Object.create( Loader.prototype ), {
```

I also had to move`setTexturePath` from legacy into the class.

**Comment by mrdoob** [comment]
> Funny- I was also testing `ObjectLoader` last night... I found that you can get it to shake without the es6 class?

```js
ObjectLoader.prototype = /*@__PURE__*/ Object.assign( Object.create( Loader.prototype ), {
```

So... I guess we didn't need to convert the codebase to es6 classes after all... 🤨

**Comment by ianpurvis** [comment]
Yeah, I didn’t want to go that far... 😆 If `/*@__PURE__*/` can really work for this, we’ll have another tool at our disposal. Want me to do a spike on that? shakediff has been very useful for testing rollup... I think it would be good to add a webpack mode to it.

**Comment by mrdoob** [comment]
Yeah... I think converting what we can to es6 classes is good, but the fact that browsers do not support private properties yet is forcing us to change the shape of the object by turning these properties to public, or turning the into symbols.

Would be great to see how things look like with `/*@__PURE__*/` if you can give it a try.

**Comment by donmccurdy** [comment]
I was wondering why ES6 Classes were related to tree-shaking, nice to know it isn't completely opaque magic. 😆

**Comment by marcofugaro** [comment]
I think it's cleaner having classes and let the bundler do the tree-shaking rather than having bundler-specific code in the codebase and check the classes' pureness manually.

BTW, `/*@__PURE__*/` is just a rollup thing, [for webpack it's `/*#__PURE__*/`](https://webpack.js.org/guides/tree-shaking/#mark-a-function-call-as-side-effect-free).

It's weird that rollup still needs that notation, I'll see if I can look into it.

**Comment by mrdoob** [comment]
It seems to me that what is enabling tree-shaking is just this pattern, right?

```js
ObjectLoader.prototype.foo = function () {}
```

Instead of this:

```js
Object.assign( ObjectLoader.prototype, {

	foo: function () {}

} );
```

**Comment by ianpurvis** [comment]
Yeah- that pattern works well, except it seems to break if you introduce inheritance like this:

```js
function ClassA() {
  return {};
}
ClassA.prototype.foo = function() {};

function ClassC() {
	return ClassA.call(this);
}
ClassC.prototype = Object.create( ClassA.prototype );
ClassC.prototype.constructor = ClassC;
ClassC.prototype.foo = function() {};
```

**Comment by ianpurvis** [comment]
I've been building up some [examples](https://github.com/ianpurvis/shakediff/commit/17f66612527d191c94208d7e69ab2994fc773176) to test tree-shaking with rollup and webpack (and also parcel now). These are helping me understand the behavior, however I will say that shaking `build/three.module.js` seems to have some issues that the isolated tests don't recreate. If anybody wants to contribute more examples, I would be glad to get them in the mix.

TLDR, es6 classes seem to be the best way to support all bundlers.

Rollup tree-shaking is the most versatile, it can shake even without es6 classes or PURE annotations. We can make a lot of progress just by replacing `Object.assign(Example.prototype, {})` with `Example.prototype.foo= {}`.

Webpack can shake a good amount of scenarios if the code is annotated, however non-es6 classes and mutations are problematic. One approach that might work would be to wrap these kind of exports in an [IIFE with a PURE annotation](https://github.com/ianpurvis/shakediff/commit/17f66612527d191c94208d7e69ab2994fc773176#diff-3a24a68cd6b24c99afdf0f15dd99c5deR8-R16) (with a build transform for instance). Not sure about the performance implications of the additional scoping, though.

Lastly, Parcel is unable to shake many situations with its vanilla config. But if I understand correctly, parcel will make use of a local `.terserrc` file, so it may be possible for people to tune their config for better shaking.

As Marco pointed out, `Three.legacy.js` is a pain point. If we distribute the remaining prototype mutations back to their target classes, following what was done for `ObjectLoader`, it will be a decent amount of work. Though I think the es6 team could knock that out fairly quickly if working together. Integrating deprecations with each class is nice because it enables them to be eliminated with their class during shaking. Plus, co-locating the deprecations with their classes should help facilitate IIFE wrapping if we have to go there.

I also like [PIXI's approach](https://pixijs.download/v5.3.3/docs/PIXI.html#.useDeprecated) to deprecations where we turn  `Three.legacy.js` into a single exported function. Something like this-
```js
import * as THREE from 'three';
THREE.enableDeprecations(); // MUST be bound to namespace
```
Since the function itself is tree-shakable, this allows elimination of all the legacy code in one shot, regardless of which classes are imported by the user. And it touches less files. 😉

**Comment by marcofugaro** [comment]
Thanks for the in-depth research @ianpurvis!

I prefer the solution in which we move all of the methods in `Three.legacy.js` in each respective class. But let's see what @mrdoob says.

Also I did some research about the PURE notation, and it turns out the approach you did in #20236 is the best one since otherwise bundler won't remove the variable and the class (if it's used only there).

**Comment by ianpurvis** [comment]
I think that could work well. @mrdoob let us know where you want to go with this, we'll hook up with the others and make it happen.
